// aspida/dist/getDirentTree.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDirentTree = void 0;
var fs_1 = __importDefault(require("fs"));
var parseInterface_1 = require("./parseInterface");
var getDirentTree = function (input) {
    var tree = { path: input, children: [] };
    fs_1.default.readdirSync(input, { withFileTypes: true })
        .filter(function (dirent) { return !dirent.name.startsWith('$') && !dirent.name.startsWith('@'); })
        .sort(function (a, b) { return (a.name < b.name ? -1 : 1); })
        .forEach(function (dirent) {
        if (dirent.isDirectory()) {
            tree.children.push({
                name: dirent.name,
                isDir: true,
                tree: exports.getDirentTree(input + "/" + dirent.name)
            });
        }
        else if (dirent.name.endsWith('.ts')) {
            var value = parseInterface_1.parse(fs_1.default.readFileSync(input + "/" + dirent.name, 'utf8'), 'Methods');
            if (!(value === null || value === void 0 ? void 0 : value.methods.some(function (_a) {
                var props = _a.props;
                return Object.keys(props).length;
            })))
                return;
            tree.children.push({
                name: dirent.name,
                isDir: false,
                methods: value.methods,
                doc: value.doc
            });
        }
    });
    return tree;
};
exports.getDirentTree = getDirentTree;
